File: DevEnvironment/Base/src/main/java/base/core/StringSearch.java
package base.core;

/**
 * The StringSearch class implements complementary APIs for string search.
 */
public final class StringSearch {
    /**
     * Gets an index of a character by searching from start.
     * Returns -1 if the character is not found.
     */
    public static int firstIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char character) {

        if (startIndex > endIndex) {
            return -1;
        }

        return firstIndexOf(str, startIndex, endIndex, character, false);
    }

    /**
     * Gets an index of a character by searching from start. (Ignore Case.)
     * Returns -1 if the character is not found.
     */
    public static int firstIndexOfIgnoreCase(
        char[] str,
        int startIndex,
        int endIndex,
        char character) {

        if (startIndex > endIndex) {
            return -1;
        }

        return firstIndexOf(str, startIndex, endIndex, character, true);
    }

    /**
     * Gets an index of a sub string by searching from start.
     * Returns -1 if the character is not found.
     */
    public static int firstIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subStr) {

        if (startIndex > endIndex) {
            return -1;
        }

        return firstIndexOf(str, startIndex, endIndex, subStr, false);
    }

    /**
     * Gets an index of a sub string by searching from start. (Ignore Case.)
     * Returns -1 if the character is not found.
     */
    public static int firstIndexOfIgnoreCase(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subStr) {

        if (startIndex > endIndex) {
            return -1;
        }

        return firstIndexOf(str, startIndex, endIndex, subStr, true);
    }

    /**
     * Gets an index of a character by searching from end.
     * Returns -1 if the character is not found.
     */
    public static int lastIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char character) {

        if (startIndex > endIndex) {
            return -1;
        }

        return lastIndexOf(str, startIndex, endIndex, character, false);
    }

    /**
     * Gets an index of a character by searching from end. (Ignore Case.)
     * Returns -1 if the character is not found.
     */
    public static int lastIndexOfIgnoreCase(
        char[] str,
        int startIndex,
        int endIndex,
        char character) {

        if (startIndex > endIndex) {
            return -1;
        }

        return lastIndexOf(str, startIndex, endIndex, character, true);
    }

    /**
     * Gets an index of a sub string by searching from end.
     * Returns -1 if the character is not found.
     */
    public static int lastIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subStr) {

        if (startIndex > endIndex) {
            return -1;
        }

        return lastIndexOf(str, startIndex, endIndex, subStr, false);
    }

    /**
     * Gets an index of a sub string by searching from end. (Ignore Case.)
     * Returns -1 if the character is not found.
     */
    public static int lastIndexOfIgnoreCase(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subString) {

        if (startIndex > endIndex) {
            return -1;
        }

        return lastIndexOf(str, startIndex, endIndex, subString, true);
    }

    /**
     * Gets an index of a character by searching from start.
     * Returns -1 if the character is not found.
     */
    private static int firstIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char character,
        boolean ignoreCase) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);

        int currIndex = startIndex;

        while (currIndex <= endIndex) {
            char currCharacter = str[currIndex];

            if (ignoreCase) {
                currCharacter = Character.toLowerCase(currCharacter);
            }

            if (currCharacter == character) {
                return currIndex;
            }

            ++currIndex;
        }

        return -1;
    }

    /**
     * Gets an index of a sub string by searching from start.
     * Returns -1 if the sub-string is not found.
     */
    private static int firstIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subStr,
        boolean ignoreCase) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);
        assert(subStr != null);

        if (subStr.length > str.length || subStr.length == 0) {
            return -1;
        }

        int currIndex = startIndex;
        int effectiveEndIndex = endIndex - subStr.length + 1;

        while (currIndex <= effectiveEndIndex) {
            if (ignoreCase) {
                if (StringEquality.equalsIgnoreCase(
                        str,
                        currIndex,
                        subStr,
                        0,
                        subStr.length - 1)) {

                    return currIndex;
                }
            } else {
                if (StringEquality.equals(
                        str,
                        currIndex,
                        subStr,
                        0,
                        subStr.length - 1)) {

                    return currIndex;
                }
            }

            ++currIndex;
        }

        return -1;
    }

    /**
     * Gets an index of a character by searching from end.
     * Returns -1 if the character is not found.
     */
    private static int lastIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char character,
        boolean ignoreCase) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);

        int currIndex = endIndex;

        while (currIndex >= startIndex) {
            char currCharacter = str[currIndex];

            if (ignoreCase) {
                currCharacter = Character.toLowerCase(currCharacter);
            }

            if (currCharacter == character) {
                return currIndex;
            }

            --currIndex;
        }

        return -1;
    }

    /**
     * Gets an index of a sub string by searching from end.
     * Returns -1 if the character is not found.
     */
    private static int lastIndexOf(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subStr,
        boolean ignoreCase) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);
        assert(subStr != null);

        int currIndex = endIndex;

        while (currIndex >= startIndex) {
            if (StringEquality.equals(
                str,
                currIndex,
                subStr,
                0,
                subStr.length,
                ignoreCase)) {

                return currIndex;
            }

            --currIndex;
        }

        return -1;
    }

    /**
     * Checks whether a string starts with a prefix.
     */
    public static boolean startsWith(char[] str, char[] postfix) {
        return startsWith(str, 0, postfix);
    }

    /**
     * Checks whether a string starts with a prefix.
     */
    public static boolean startsWith(char[] str, int startIndex, char[] prefix) {
        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(prefix != null);

        int strLength = str.length - startIndex;

        if (strLength < prefix.length) {
            return false;
        }

        int counter = 0;

        int strIndex = startIndex;
        int prefixIndex = 0;

        while (counter < prefix.length && strIndex < str.length && prefixIndex < prefix.length) {
            char strCharacter = str[strIndex];
            char prefixCharacter = prefix[prefixIndex];

            if (strCharacter != prefixCharacter) {
                return false;
            }

            ++counter;

            ++strIndex;
            ++prefixIndex;
        }

        return strIndex == startIndex + counter;
    }

    /**
     * Checks whether the string ends with a postfix.
     */
    public static boolean endsWith(char[] str, char[] postfix) {
        return endsWith(str, str.length - 1, postfix);
    }

    /**
     * Checks whether the string ends with a postfix.
     */
    public static boolean endsWith(char[] str, int endIndex, char[] postfix) {
        assert(str != null);
        assert(endIndex >= 0 && endIndex < str.length);
        assert(postfix != null);

        int strLength = Indexes.size(0, endIndex);

        if (strLength < postfix.length) {
            return false;
        }

        int counter = 0;

        int strIndex = endIndex;
        int postfixIndex = postfix.length - 1;

        while (counter < postfix.length && strIndex >= 0 && postfixIndex >= 0) {
            char strCharacter = str[strIndex];
            char postfixCharacter = postfix[postfixIndex];

            if (strCharacter != postfixCharacter) {
                return false;
            }

            ++counter;

            --strIndex;
            --postfixIndex;
        }

        return strIndex == endIndex - counter;
    }

    /**
     * Counts the number of instances of a substring.
     */
    public static int countSubString(
        String str,
        String subString) {

        boolean allowOverlapping = true;
        return countSubString(str, subString, allowOverlapping);
    }

    /**
     * Counts the number of instances of a substring.
     */
    public static int countSubString(
        String str,
        String subString,
        boolean allowOverlapping) {

        if (str == null || str.isEmpty() || subString == null || subString.isEmpty()) {
            return 0;
        }

        int startIndex = 0;
        int endIndex = str.length() - 1;
        int subStringStartIndex = 0;
        int subStringEndIndex = subString.length() - 1;

        return countSubString(
            str,
            startIndex,
            endIndex,
            subString,
            subStringStartIndex,
            subStringEndIndex,
            allowOverlapping);
    }

    /**
     * Counts the number of instances of a substring.
     */
    public static int countSubString(
        String str,
        int startIndex,
        int endIndex,
        String subString,
        int subStringStartIndex,
        int subStringEndIndex,
        boolean allowOverlapping) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length());
        assert(endIndex >= startIndex && endIndex < str.length());

        assert(subString != null);
        assert(subStringStartIndex >= 0 && subStringStartIndex < subString.length());
        assert(subStringEndIndex >= subStringStartIndex && subStringEndIndex < subString.length());

        return countSubString(
            str.toCharArray(),
            startIndex,
            endIndex,
            subString.toCharArray(),
            subStringStartIndex,
            subStringEndIndex,
            allowOverlapping);
    }

    /**
     * Counts the number of instances of a substring.
     */
    public static int countSubString(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subString,
        int subStringStartIndex,
        int subStringEndIndex,
        boolean allowOverlapping) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);

        assert(subString != null);
        assert(subStringStartIndex >= 0 && subStringStartIndex < subString.length);
        assert(subStringEndIndex >= subStringStartIndex && subStringEndIndex < subString.length);

        int stringLength = Indexes.size(startIndex, endIndex);
        int subStringLength = Indexes.size(subStringStartIndex, subStringEndIndex);

        if (subStringLength == 0 || subStringLength > stringLength) {
            return 0;
        }

        int counter = 0;

        int currIndex = startIndex;
        int effectiveEndIndex = endIndex - subStringLength + 1;

        while (currIndex <= effectiveEndIndex) {
            if (StringEquality.equals(str, currIndex, subString, subStringStartIndex, subStringLength)) {
                ++counter;

                currIndex = (allowOverlapping) ? currIndex + 1 : currIndex + subStringLength;
            }
            else {
                ++currIndex;
            }
        }

        return counter;
    }

    /**
     * Counts the number of instances of a substring from end.
     * Avoid overlapping.
     */
    public static int countSubStringFromEnd(
        char[] str,
        int startIndex,
        int endIndex,
        char[] subString,
        int subStringStartIndex,
        int subStringEndIndex,
        boolean allowOverlapping) {

        assert(str != null);
        assert(startIndex >= 0 && startIndex < str.length);
        assert(endIndex >= startIndex && endIndex < str.length);

        assert(subString != null);
        assert(subStringStartIndex >= 0 && subStringStartIndex < subString.length);
        assert(subStringEndIndex >= subStringStartIndex && subStringEndIndex < subString.length);

        int stringLength = Indexes.size(startIndex, endIndex);
        int subStringLength = Indexes.size(subStringStartIndex, subStringEndIndex);

        if (subStringLength == 0 || subStringLength > stringLength) {
            return 0;
        }

        int counter = 0;

        int currIndex = endIndex - subStringLength + 1;

        while (currIndex >= startIndex) {
            if (StringEquality.equals(str, currIndex, subString, subStringStartIndex, subStringLength)) {
                ++counter;

                currIndex = (allowOverlapping) ? currIndex - 1 : currIndex - subStringLength;
            }
            else {
                --currIndex;
            }
        }

        return counter;
    }

    /**
     * The StringSearch constructor - Disables the default constructor.
     */
    private StringSearch() {
    }
}

File: DevEnvironmentCLion/Base/core/StringSearch.cpp
#include "PreCompiled.h"
#include "StringSearch.h"
#include "StringEquality.h"
#include "Dimensions.h"

using namespace base;

/**
 * The CaseInsensitiveCompare class implements a case insensitive character comparison.
 */
class CaseInsensitiveCompare : public std::binary_function<char, char, bool>
{
public:
    bool operator()(char left, char right) const
    {
        return std::tolower(left) == std::tolower(right);
    }
};

/**
 * Gets first index of a character.
 * Returns string::npos if the character is not found.
 */
size_t StringSearch::firstIndexOf(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    char character)
{
    if (startIndex > endIndex)
    {
        return std::string::npos;
    }

    std::string stringToSearch(1, character);
    size_t length = Dimensions::indexes(startIndex, endIndex);

    std::string::const_iterator resultIterator = std::find_first_of(
        str.begin(),
        str.begin() + length,
        stringToSearch.begin(),
        stringToSearch.end());

    if (resultIterator == str.end())
    {
        return std::string::npos;
    }

    return resultIterator - str.begin();
}

/**
 * Gets first index of a character ignore case. (Ignore Case)
 * Returns string::npos if the character is not found.
 */
size_t StringSearch::firstIndexOfIgnoreCase(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    char character)
{
    if (startIndex > endIndex)
    {
        return std::string::npos;
    }

    std::string stringToSearch(1, character);
    size_t length = Dimensions::indexes(startIndex, endIndex);

    std::string::const_iterator resultIterator = std::find_first_of(
        str.begin(),
        str.begin() + length,
        stringToSearch.begin(),
        stringToSearch.end(),
        CaseInsensitiveCompare());

    if (resultIterator == str.end())
    {
        return std::string::npos;
    }

    return resultIterator - str.begin();
}

/**
 * Gets last index of a character.
 * Returns string::npos if the character is not found.
 */
size_t StringSearch::lastIndexOf(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    char character)
{
    if (startIndex > endIndex)
    {
        return std::string::npos;
    }

    std::string stringToSearch(1, character);
    size_t length = endIndex - startIndex + 1;

    std::string::const_iterator resultIterator = std::find_end(
        str.begin(),
        str.begin() + length,
        stringToSearch.begin(),
        stringToSearch.end());

    if (resultIterator == str.end())
    {
        return std::string::npos;
    }

    return resultIterator - str.begin();
}

/**
* Gets last index of a character. (Ignore Case)
* Returns string::npos if the character is not found.
*/
size_t StringSearch::lastIndexOfIgnoreCase(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    char character)
{
    if (startIndex > endIndex)
    {
        return std::string::npos;
    }

    std::string stringToSearch(1, character);
    size_t length = endIndex - startIndex + 1;

    std::string::const_iterator resultIterator = std::find_end(
        str.begin(),
        str.begin() + length,
        stringToSearch.begin(),
        stringToSearch.end(),
        CaseInsensitiveCompare());

    if (resultIterator == str.end())
    {
        return std::string::npos;
    }

    return resultIterator - str.begin();
}

/**
 * Checks whether a string starts with a prefix.
 */
bool StringSearch::startsWith(
    const std::string& str,
    const std::string& prefix)
{
    return startsWith(str, 0, prefix);
}

/**
 * Checks whether a string starts with a prefix.
 */
bool StringSearch::startsWith(
    const std::string& str,
    size_t startIndex,
    const std::string& prefix)
{
    if (str.empty() || prefix.empty() || prefix.size() > str.size() - startIndex)
    {
        return std::string::npos;
    }

    size_t prefixStartIndex = str.find(prefix.c_str(), startIndex);
    return prefixStartIndex != std::string::npos && prefixStartIndex == startIndex;
}

/**
 * Checks whether a string ends with a postfix.
 */
bool StringSearch::endsWith(
    const std::string& str,
    const std::string& postfix)
{
    if (str.empty())
    {
        return std::string::npos;
    }

    return endsWith(str, str.size() - 1, postfix);
}

/**
 * Checks whether a string ends with a postfix.
 */
bool StringSearch::endsWith(
    const std::string& str,
    size_t endIndex,
    const std::string& postfix)
{
    if (str.empty() || postfix.empty() || postfix.size() > endIndex + 1)
    {
        return std::string::npos;
    }

    size_t postfixStartIndex = str.rfind(postfix.c_str(), endIndex);
    return postfixStartIndex != std::string::npos && postfixStartIndex == endIndex - postfix.size() + 1;
}

/**
 * Counts the number of instances of a substring.
 */
unsigned int StringSearch::countSubString(
    const std::string& str,
    const std::string& subString)
{
    bool allowOverlapping = true;
    return countSubString(str, subString, allowOverlapping);
}

/**
 * Counts the number of instances of a substring.
 */
unsigned int StringSearch::countSubString(
    const std::string& str,
    const std::string& subString,
    bool allowOverlapping)
{
    if (str.empty() || subString.empty())
    {
        return 0;
    }

    size_t startIndex = 0;
    size_t endIndex = str.length() - 1;
    size_t subStringStartIndex = 0;
    size_t subStringEndIndex = subString.length() - 1;

    return countSubString(
        str,
        startIndex,
        endIndex,
        subString,
        subStringStartIndex,
        subStringEndIndex,
        allowOverlapping);
}

/**
 * Counts the number of instances of a substring.
 */
unsigned int StringSearch::countSubString(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    const std::string& subString,
    size_t subStringStartIndex,
    size_t subStringEndIndex,
    bool allowOverlapping)
{
    assert(startIndex >= 0 && startIndex < str.size());
    assert(endIndex >= startIndex && endIndex < str.size());

    assert(subStringStartIndex >= 0 && subStringStartIndex < subString.size());
    assert(subStringEndIndex >= subStringStartIndex && subStringEndIndex < subString.size());

    size_t stringLength = Dimensions::indexes(startIndex, endIndex);
    size_t subStringLength = Dimensions::indexes(subStringStartIndex, subStringEndIndex);

    if (subStringLength == 0 || subStringLength > stringLength) {
        return 0;
    }

    unsigned int counter = 0;

    size_t currIndex = startIndex;
    size_t effectiveEndIndex = endIndex - subStringLength + 1;

    while (currIndex <= effectiveEndIndex) {
        if (StringEquality::equals(str, currIndex, subString, subStringStartIndex, subStringLength)) {
            ++counter;

            currIndex = (allowOverlapping) ? currIndex + 1 : currIndex + subStringLength;
        }
        else {
            ++currIndex;
        }
    }

    return counter;
}

/**
 * Counts the number of instances of a substring from end.
 * Avoid overlapping.
 */
unsigned int StringSearch::countSubStringFromEnd(
    const std::string& str,
    size_t startIndex,
    size_t endIndex,
    const std::string& subString,
    size_t subStringStartIndex,
    size_t subStringEndIndex,
    bool allowOverlapping)
{
    assert(startIndex >= 0 && startIndex < str.size());
    assert(endIndex >= startIndex && endIndex < str.size());

    assert(subStringStartIndex >= 0 && subStringStartIndex < subString.size());
    assert(subStringEndIndex >= subStringStartIndex && subStringEndIndex < subString.size());

    size_t stringLength = Dimensions::indexes(startIndex, endIndex);
    size_t subStringLength = Dimensions::indexes(subStringStartIndex, subStringEndIndex);

    if (subStringLength == 0 || subStringLength > stringLength) {
        return 0;
    }

    unsigned int counter = 0;

    size_t currIndex = endIndex - subStringLength + 1;

    while (currIndex >= startIndex) {
        if (StringEquality::equals(str, currIndex, subString, subStringStartIndex, subStringLength)) {
            ++counter;

            currIndex = (allowOverlapping) ? currIndex - 1 : currIndex - subStringLength;
        }
        else {
            --currIndex;
        }
    }

    return counter;
}
